### PR TITLE
Add a role to deploy Keycloak

### PR DESCRIPTION
**Pull request by m-bull**
Deploy Keycloak using the official OCI container image and the JDBC_PING [JGroups](http://jgroups.org/) discovery protocol.

### CODE REVIEW

File: roles/deploy/README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
+This role deploys Keycloak in standalone-HA configuration, using MariaDB as the backend storage 
+
+### Role variables
+
+#### OCI image
+##### Optional
+* **keycloak_docker_name:** Name of the Keycloak docker container (default: "keycloak")
+* **keycloak_docker_image:** Docker image to use (default: "quay.io/keycloak/keycloak")
+* **keycloak_docker_tag:** Image tag to use (default: "21.1")
+ **keycloak_container.etc_hosts:**
```
Comment (markgoddard) [inline roles/deploy/README.md]:
```suggestion
*  **keycloak_container.etc_hosts:** Dict of host-to-IP mappings
```

File: roles/deploy/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    owner: root
+    mode: 0644
+  become: true
+
+- name: Template Keycloak cache ISPN configuration
+  template:
+    src: "{{ keycloak_cache_config_file_template }}"
+    dest: "{{ keycloak_host_config_dir }}/{{ keycloak_cache_config_file }}"
+    owner: root
+    mode: 0644
```
Comment (markgoddard) [inline roles/deploy/tasks/main.yml:16]: Stricter permissions required here - it has the DB password in

File: roles/deploy/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    image: "{{ keycloak_docker_image }}:{{ keycloak_docker_tag }}"
+    network_mode: host
+    etc_hosts: "{{ keycloak_container.etc_hosts | default(omit) }}"
+    volumes: "{{ _keycloak_volumes }}"
+    comparisons:
+      '*': strict
+    restart_policy: "always"
+    env:
+      KEYCLOAK_ADMIN: "{{ keycloak_admin_username }}"
+      KEYCLOAK_ADMIN_PASSWORD: "{{ keycloak_admin_password }}"
```
Comment (markgoddard) [inline roles/deploy/tasks/main.yml:32]: These two aren't in defaults or README

File: roles/deploy/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      KC_METRICS_ENABLED: "{{ keycloak_metrics_enabled | default('false') | string }}"
+      KC_HEALTH_ENABLED: "{{ keycloak_health_enabled | default('false') | string }}"
+      KC_CACHE_CONFIG_FILE: "{{ keycloak_cache_config_file }}"
+      KC_JDBC_DRIVER_NAME: "{{ keycloak_jdbc_driver_name }}"
+      KC_JDBC_DRIVER_INITIALISE_SQL: "{{ keycloak_jdbc_initialise_sql }}"
+      JGROUPS_DISCOVERY_PROTOCOL: "{{ keycloak_jgroups_discovery_protocol }}"
+      JGROUPS_DISCOVERY_BIND_ADDR: "{{ keycloak_jgroups_discovery_bind_addr }}"
+      JGROUPS_DISCOVERY_BIND_PORT: "{{ keycloak_jgroups_discovery_bind_port | string }}"
+    command: "{{ keycloak_docker_command }}"
+  register: ensure_keycloak_container
```
Comment (markgoddard) [inline roles/deploy/tasks/main.yml]: Unused

File: roles/deploy/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      KC_DB_URL_PORT: "{{ keycloak_db_url_port | string }}"
+      KC_DB_URL_DATABASE: "{{ keycloak_db_url_database }}"
+      KC_HOSTNAME_URL: "{{ keycloak_hostname_url | default(omit) }}"
+      KC_HTTP_HOST: "{{ keycloak_http_host | default(omit) }}"
+      KC_HTTP_PORT: "{{ keycloak_http_port | default(omit) }}"
+      KC_HOSTNAME_STRICT: "{{ keycloak_hostname_strict | default(omit) }}"
+      KC_HOSTNAME_PATH: "{{ keycloak_hostname_path | default(omit) }}"
+      KC_HTTP_RELATIVE_PATH: "{{ keycloak_http_relative_path | default(omit) }}"
+      KC_METRICS_ENABLED: "{{ keycloak_metrics_enabled | default('false') | string }}"
+      KC_HEALTH_ENABLED: "{{ keycloak_health_enabled | default('false') | string }}"
```
Comment (markgoddard) [inline roles/deploy/tasks/main.yml:47]: Not in defaults or README

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: roles/deploy/README.md (hunk 1 above)
Comment (m-bull) [reply roles/deploy/README.md]: Updated

File: roles/deploy/tasks/main.yml (hunk 3 above)
Comment (m-bull) [reply roles/deploy/tasks/main.yml:32]: Updated

File: roles/deploy/tasks/main.yml (hunk 4 above)
Comment (m-bull) [reply roles/deploy/tasks/main.yml]: Removed

File: roles/deploy/tasks/main.yml (hunk 5 above)
Comment (m-bull) [reply roles/deploy/tasks/main.yml:47]: Added to both

File: roles/deploy/tasks/main.yml (hunk 2 above)
Comment (m-bull) [reply roles/deploy/tasks/main.yml:16]: Nothing actually gets exposed in this file (which is quite nice!). There's functionality in the container for filling in `"${env.KC_DB_PASSWORD}"` in the xml file with an environment variable at runtime.

Comment (markgoddard) [reply roles/deploy/tasks/main.yml:16]: Ack, makes sense.

**Review by markgoddard — APPROVED** [review]
Nice work

**Pull request merged**